code/tools/QgenMain.c

```diff
@@ -68,7 +68,7 @@ int g_nQueryNumber,
 StringBuffer_t *g_sbTemplateName = NULL;
 
 
-int yydebug;
+extern int yydebug;
```

Comment:
This change causes compilation failure on my side. Could you revert this change and see if it compiles on your side?

Reply:
I tried reverting the change; without externing `yydebug`, I get a linker error when trying to compile `dsqgen`:

```
$ make
...
gcc -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DYYDEBUG  -DLINUX -g -Wall  -o dsqgen address.o date.o decimal.o dist.o error_msg.o expr.o eval.o genrand.o grammar_support.o keywords.o list.o nulls.o print.o permute.o QgenMain.o query_handler.o r_params.o scaling.o StringBuffer.o substitution.o tdefs.o text.o tokenizer.o w_inventory.o y.tab.o release.o scd.o build_support.o parallel.o  utility.o -lm
/usr/bin/ld: y.tab.o:/home/wanshenl/git/dsb/code/tools/y.tab.c:1245: multiple definition of `yydebug'; QgenMain.o:/home/wanshenl/git/dsb/code/tools/QgenMain.c:71: first defined here
collect2: error: ld returned 1 exit status
make: *** [makefile:235: dsqgen] Error 1
```

My environment is the following:
```
$ lsb_release -cdr
Description:    Ubuntu 22.04 LTS
Release:        22.04
Codename:       jammy
$ uname -r
5.15.0-39-generic
$ gcc --version
gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
$ bison --version
bison (GNU Bison) 3.8.2
$ flex --version
flex 2.6.4
```

Hmm. I will try to look for a workaround.

Reply:
I found a workaround that involves specifying `-fcommon` as a `gcc` compiler flag. The details are in the commit message for the last commit (`42896f5`), essentially, `gcc-10` changed default behavior for whether the linker can merge global variables with the same name.

I wonder if you will have similar compilation errors for the other variables that I externed,
if so, please let me know and I will remove those `extern`s too.

I tested a bit further, just the addition of `-fcommon` is enough to make all of the initial code compile on my environment.
However, while `-fcommon` is safe to use for `yydebug` since there should only be one instance of that variable,
the other variables that I renamed and externed may not be safe to combine as they seem to be conceptually distinct.
For example, `pDemoRecord` was originally defined in two ways:

```
// w_customer_demographics.c
                int nMaxAttribute = distsize("gender") * distsize("education") * distsize("marital_status");
		int nKey = (int)get_rowcount(CUSTOMER_DEMOGRAPHICS);
		pDemoRecord = initializeCharAttributeKeyRecord(nMaxAttribute, nKey);

// w_household_demographics.c
                int nMaxAttribute = distsize("buy_potential") * distsize("dependent_count");
		int nKey = (int)get_rowcount(HOUSEHOLD_DEMOGRAPHICS);
		pDemoRecord = initializeCharAttributeKeyRecord(nMaxAttribute, nKey);
```

and it seems like we would not want to merge these definitions together.

So I think we would want to (1) specify `-fcommon` and (2) keep the variable renaming from this PR.

Thanks!

Reply:
> I tried reverting the change; without externing `yydebug`, I get a linker error when trying to compile `dsqgen`:
> 
> ```
> $ make
> ...
> gcc -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -DYYDEBUG  -DLINUX -g -Wall  -o dsqgen address.o date.o decimal.o dist.o error_msg.o expr.o eval.o genrand.o grammar_support.o keywords.o list.o nulls.o print.o permute.o QgenMain.o query_handler.o r_params.o scaling.o StringBuffer.o substitution.o tdefs.o text.o tokenizer.o w_inventory.o y.tab.o release.o scd.o build_support.o parallel.o  utility.o -lm
> /usr/bin/ld: y.tab.o:/<snip>/dsb/code/tools/y.tab.c:1245: multiple definition of `yydebug'; QgenMain.o:/<snip>/dsb/code/tools/QgenMain.c:71: first defined here
> collect2: error: ld returned 1 exit status
> make: *** [makefile:235: dsqgen] Error 1
> ```
> 
> My environment is the following:
> 
> ```
> $ lsb_release -cdr
> Description:    Ubuntu 22.04 LTS
> Release:        22.04
> Codename:       jammy
> $ uname -r
> 5.15.0-39-generic
> $ gcc --version
> gcc (Ubuntu 11.2.0-19ubuntu1) 11.2.0
> $ bison --version
> bison (GNU Bison) 3.8.2
> $ flex --version
> flex 2.6.4
> ```
> 
> Hmm. I will try to look for a workaround.

Hi Wan, could you check if yydebug is defined elsewhere in your local repo where the change is not included in the PR? Maybe check if the yydebug is defined in y.tab.c in your local repo, and try to revert that as well and see if it can compile without the linking error.

Reply:
Hi Bailu,

As of commit 5a5c3bcaf5f501019203c2c498e33bc746431fed , `git status` returns a clean work tree. From there,

`yydebug` is already defined by DSB: https://github.com/microsoft/dsb/blob/main/code/tools/y.tab.c#L1117
And this definition is active because our CFLAGS include `-DYYDEBUG`, https://github.com/microsoft/dsb/blob/main/code/tools/makefile#L62

Even if I hack the makefile so that it does not regenerate `y.tab.c` (by no-op-ing https://github.com/microsoft/dsb/blob/main/code/tools/makefile#L283-L284),
it will have a link error.

Could you please check if your local environment somehow not including `-DYYDEBUG`?
e.g., maybe you are using an IDE that overrides CFLAGS. I am running directly from the terminal.

---

That said, the latest commit 42896f5647c2da79961d24f634843e603d699729 should have unified behavior across old and new GCC versions, does it still not compile?

Reply:
Hi @lmwnshn 

Thank you for providing the fix!

I also ran into the same problem when trying to run the makefile.

Using your branch at the latest commit, I was able to compile everything without errors.
In case it still helps, the environment I used was the following:

* Ubuntu 22.04.3 LTS
* 6.2.0-33-generic kernel
* gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
* bison (GNU Bison) 3.8.2
* flex 2.6.4

Reply:
@arselzer Good to know, thanks for the info. I will link from the issue in case anyone else finds it helpful.

Reply:
Hi @lmwnshn，
大佬真是救大命了What you do really helps me!! Thanks capybara